pkg/runit/servicebuilder.go

```diff
@@ -18,51 +18,85 @@
 package runit
 
 import (
-	"bytes"
 	"fmt"
-	"io"
+	"io/ioutil"
 	"os"
-	"os/exec"
-	"path"
+	"path/filepath"
 
 	"github.com/square/p2/pkg/util"
 
 	"gopkg.in/yaml.v2"
 )
 
-type SBTemplate struct {
-	name       string
-	runClauses map[string]map[string][]string
+type ServiceTemplate struct {
+	Run      []string `yaml:"run"`
+	Log      []string `yaml:"log,omitempty"`
+	Sleep    int      `yaml:"sleep,omitempty"`
+	LogSleep int      `yaml:"logsleep,omitempty"`
 }
 
-type SBTemplateEntry struct {
-	baseCmd []string
-}
+func (s ServiceTemplate) RunScript() ([]byte, error) {
+	if len(s.Run) == 0 {
+		return nil, util.Errorf("empty run script")
+	}
+
+	args, err := yaml.Marshal(s.Run)
+	if err != nil {
+		return nil, err
+	}
 
-func NewSBTemplate(name string) *SBTemplate {
-	return &SBTemplate{
-		name:       name,
-		runClauses: make(map[string]map[string][]string),
+	// default sleep to reduce spinning on a broken run script
+	sleep := 2
+	if s.Sleep != 0 {
+		sleep = s.Sleep
 	}
-}
 
-func (template *SBTemplate) AddEntry(app string, run []string) {
-	template.runClauses[app] = map[string][]string{"run": run}
+	ret := fmt.Sprintf(`#!/usr/bin/ruby
+$stderr.reopen(STDOUT)
+require 'yaml'
+sleep %v
+exec *YAML.load(DATA.read)
+__END__
+%s`, sleep, args)
+	return []byte(ret), nil
 }
 
-func (template *SBTemplate) Write(out io.Writer) error {
-	b, err := yaml.Marshal(template.runClauses)
+func (s ServiceTemplate) LogScript() ([]byte, error) {
+	if len(s.Log) == 0 {
+		// use a default log script that makes a logdir, chowns it and execs
+		// svlogd into it
+		return []byte(`#!/usr/bin/ruby
+require 'fileutils'
+FileUtils.mkdir_p('./main')
+FileUtils.chown('nobody', 'nobody', './main')
+sleep 2
+exec('chpst', '-unobody', 'svlogd', '-tt', './main')
```

Comment:
used chpst because that's what our old log script was doing and I didn't see much reason to change it. we could probably use p2-exec instead, I'll take a look. I would rather not introduce configurable paths for svlogd just because I don't think there's any practical reason to do so and passing those paths down the structs is a pain.